File: src/data_in_ai_revolution/doctor.py
from __future__ import annotations

from dataclasses import dataclass
import importlib.util
from pathlib import Path
import shutil
import subprocess
import sys
from typing import Callable

from .catalog import discover_repo_root


PackageChecker = Callable[[str], bool]
CommandChecker = Callable[[str], bool]
GitRunner = Callable[[Path], tuple[str, str]]


@dataclass(slots=True)
class DoctorCheck:
    name: str
    status: str
    detail: str
    hint: str = ""

    @property
    def icon(self) -> str:
        return {"pass": "✓", "warn": "!", "fail": "✗"}.get(self.status, "•")


CORE_PACKAGES = ("textual", "rich")
LAB_PACKAGES = ("jupyter", "torch", "transformers", "matplotlib", "sentence_transformers", "manim")
RECOMMENDED_COMMANDS = ("jupyter", "ffmpeg")


def default_package_checker(package: str) -> bool:
    return importlib.util.find_spec(package) is not None


def default_command_checker(command: str) -> bool:
    return shutil.which(command) is not None


def default_git_runner(repo_root: Path) -> tuple[str, str]:
    branch = subprocess.run(
        ["git", "branch", "--show-current"],
        cwd=repo_root,
        check=False,
        capture_output=True,
        text=True,
    ).stdout.strip() or "detached"
    dirty = subprocess.run(
        ["git", "status", "--short"],
        cwd=repo_root,
        check=False,
        capture_output=True,
        text=True,
    ).stdout.strip()
    return branch, dirty


def run_doctor(
    repo_root: Path | None = None,
    *,
    package_checker: PackageChecker = default_package_checker,
    command_checker: CommandChecker = default_command_checker,
    git_runner: GitRunner = default_git_runner,
) -> tuple[DoctorCheck, ...]:
    root = discover_repo_root(repo_root)
    checks: list[DoctorCheck] = []

    readme_exists = (root / "README.md").exists()
    checks.append(
        DoctorCheck(
            "Repository layout",
            "pass" if readme_exists and (root / "notebooks").is_dir() and (root / "scripts").is_dir() else "fail",
            f"Root: {root}",
            "Expected README.md, notebooks/, and scripts/ to exist.",
        )
    )

    checks.append(
        DoctorCheck(
            "Python runtime",
            "pass" if sys.version_info >= (3, 10) else "warn",
            f"Python {sys.version.split()[0]}",
            "Use Python 3.10 or newer for the TUI and test suite.",
        )
    )

    notebook_count = len(list((root / "notebooks").glob("*.ipynb")))
    checks.append(
        DoctorCheck(
            "Workshop notebooks",
            "pass" if notebook_count else "warn",
            f"Found {notebook_count} notebook(s)",
            "Notebooks power the lab launcher and hands-on sections.",
        )
    )

    core_installed = [package for package in CORE_PACKAGES if package_checker(package)]
    checks.append(
        DoctorCheck(
            "Core TUI packages",
            "pass" if len(core_installed) == len(CORE_PACKAGES) else "fail",
            f"Installed: {', '.join(core_installed) if core_installed else 'none'}",
            "Run `python -m pip install -e .` to install the terminal lab.",
        )
    )

    available_lab_packages = [package for package in LAB_PACKAGES if package_checker(package)]
    checks.append(
        DoctorCheck(
            "Lab dependencies",
            "pass" if len(available_lab_packages) == len(LAB_PACKAGES) else "warn",
            f"Installed: {', '.join(available_lab_packages) if available_lab_packages else 'none'}",
            "Optional, but useful: `python -m pip install -r scripts/requirements.txt`.",
        )
    )

    available_commands = [command for command in RECOMMENDED_COMMANDS if command_checker(command)]
    checks.append(
        DoctorCheck(
            "External commands",
            "pass" if len(available_commands) == len(RECOMMENDED_COMMANDS) else "warn",
            f"Available: {', '.join(available_commands) if available_commands else 'none'}",
            "Jupyter launches notebooks, ffmpeg helps with video workflows.",
        )
    )

    branch, dirty = git_runner(root)
    checks.append(
        DoctorCheck(
            "Git status",
            "pass" if not dirty else "warn",
            f"Branch: {branch}; dirty files: {len(dirty.splitlines()) if dirty else 0}",
            "A clean branch makes screenshot generation and reviews easier.",
        )
    )

    return tuple(checks)


def doctor_markdown(checks: tuple[DoctorCheck, ...]) -> str:
    lines = ["# Environment doctor", ""]
    for check in checks:
        line = f"- {check.icon} **{check.name}**: {check.detail}"
        if check.hint:
            line += f" ({check.hint})"
        lines.append(line)
    return "\n".join(lines)

File: internal/doctor/doctor.go
package doctor

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
)

type Check struct {
	Name   string `json:"name"`
	Status string `json:"status"`
	Detail string `json:"detail"`
	Hint   string `json:"hint,omitempty"`
}

func (c Check) Icon() string {
	switch c.Status {
	case "pass":
		return "✓"
	case "warn":
		return "!"
	case "fail":
		return "✗"
	default:
		return "•"
	}
}

func Run(root string) []Check {
	checks := []Check{}
	checks = append(checks, Check{
		Name:   "Repository layout",
		Status: ternary(fileExists(filepath.Join(root, "README.md")) && dirExists(filepath.Join(root, "notebooks")) && dirExists(filepath.Join(root, "scripts")), "pass", "fail"),
		Detail: fmt.Sprintf("Root: %s", root),
		Hint:   "Expected README.md, notebooks/, and scripts/ to exist.",
	})

	checks = append(checks, Check{
		Name:   "Go runtime",
		Status: "pass",
		Detail: runtime.Version(),
		Hint:   "Bubble Tea runs entirely on the Go toolchain.",
	})

	pythonVersion, pythonOK := versionOutput("python3", "--version")
	checks = append(checks, Check{
		Name:   "Python runtime",
		Status: ternary(pythonOK, "pass", "warn"),
		Detail: valueOrNone(pythonVersion),
		Hint:   "Python is still useful for notebooks and workshop scripts.",
	})

	notebookCount := countGlob(filepath.Join(root, "notebooks", "*.ipynb"))
	checks = append(checks, Check{
		Name:   "Workshop notebooks",
		Status: ternary(notebookCount > 0, "pass", "warn"),
		Detail: fmt.Sprintf("Found %d notebook(s)", notebookCount),
		Hint:   "Notebooks power the labs launcher and hands-on sections.",
	})

	availableCommands := []string{}
	for _, command := range []string{"go", "python3", "jupyter", "ffmpeg"} {
		if _, err := exec.LookPath(command); err == nil {
			availableCommands = append(availableCommands, command)
		}
	}
	commandStatus := "warn"
	if len(availableCommands) >= 3 {
		commandStatus = "pass"
	}
	checks = append(checks, Check{
		Name:   "External commands",
		Status: commandStatus,
		Detail: fmt.Sprintf("Available: %s", strings.Join(availableCommands, ", ")),
		Hint:   "Go runs the TUI, Python and Jupyter launch labs, ffmpeg supports video workflows.",
	})

	branch, dirtyCount := gitStatus(root)
	checks = append(checks, Check{
		Name:   "Git status",
		Status: ternary(dirtyCount == 0, "pass", "warn"),
		Detail: fmt.Sprintf("Branch: %s; dirty files: %d", branch, dirtyCount),
		Hint:   "A clean branch keeps screenshots, reviews, and merges predictable.",
	})

	return checks
}

func Markdown(checks []Check) string {
	lines := []string{"# Environment doctor", ""}
	for _, check := range checks {
		line := fmt.Sprintf("- %s **%s**: %s", check.Icon(), check.Name, check.Detail)
		if check.Hint != "" {
			line += fmt.Sprintf(" (%s)", check.Hint)
		}
		lines = append(lines, line)
	}
	return strings.Join(lines, "\n")
}

func JSON(checks []Check) ([]byte, error) {
	return json.MarshalIndent(checks, "", "  ")
}

func dirExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}

func fileExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && !info.IsDir()
}

func versionOutput(command string, args ...string) (string, bool) {
	cmd := exec.Command(command, args...)
	output, err := cmd.CombinedOutput()
	if err != nil {
		return strings.TrimSpace(string(output)), false
	}
	return strings.TrimSpace(string(output)), true
}

func valueOrNone(value string) string {
	if value == "" {
		return "none"
	}
	return value
}

func countGlob(pattern string) int {
	matches, err := filepath.Glob(pattern)
	if err != nil {
		return 0
	}
	return len(matches)
}

func gitStatus(root string) (string, int) {
	branch := "detached"
	if output, err := exec.Command("git", "-C", root, "branch", "--show-current").Output(); err == nil {
		trimmed := strings.TrimSpace(string(output))
		if trimmed != "" {
			branch = trimmed
		}
	}
	if output, err := exec.Command("git", "-C", root, "status", "--short").Output(); err == nil {
		trimmed := strings.TrimSpace(string(output))
		if trimmed == "" {
			return branch, 0
		}
		return branch, len(strings.Split(trimmed, "\n"))
	}
	return branch, 0
}

func ternary[T any](condition bool, yes, no T) T {
	if condition {
		return yes
	}
	return no
}
